src/pint/gridutils.py

```diff
@@ -1,16 +1,400 @@
+    ncpu : int, optional
+        If an existing Executor is not supplied, one will be created with this number of workers.
+        If 1, will run single-processor version
+        If None, will use :func:`multiprocessing.cpu_count`
+    chunksize : int
+        Size of the chunks for :class:`concurrent.futures.ProcessPoolExecutor` parallel execution.
+        Ignored for :class:`concurrent.futures.ThreadPoolExecutor`
+    printprogress : bool, optional
+        Print indications of progress (requires :mod:`tqdm`)
+
+    Returns
+    -------
+    np.ndarray : array of chisq values
+
+    Example
+    -------
+    >>> import astropy.units as u
+    >>> import numpy as np
+    >>> import pint.config
+    >>> import pint.gridutils
+    >>> from pint.fitter import WLSFitter
+    >>> from pint.models.model_builder import get_model, get_model_and_toas
+    # Load in a basic dataset
+    >>> parfile = pint.config.examplefile("NGC6440E.par")
+    >>> timfile = pint.config.examplefile("NGC6440E.tim")
+    >>> m, t = get_model_and_toas(parfile, timfile)
+    >>> f = WLSFitter(t, m)
+    # find the best-fit
+    >>> f.fit_toas()
+    >>> bestfit = f.resids.chi2
+    # We'll do something like 3-sigma around the best-fit values of  F0
+    >>> F0 = np.linspace(f.model.F0.quantity - 3 * f.model.F0.uncertainty,f.model.F0.quantity + 3 * f.model.F0.uncertainty,25)
+    >>> chi2_F0 = pint.gridutils.grid_chisq(f, ("F0",), (F0,))
+
+    Notes
+    -----
+    By default, it will create :class:`~concurrent.futures.ProcessPoolExecutor`
+    with ``max_workers`` equal to the desired number of cpus.
+    However, if you are running this as a script you may need something like::
+
+        import multiprocessing
+
+        if __name__ == "__main__":
+            multiprocessing.freeze_support()
+            ...
+            grid_chisq(...)
+
+    If an instantiated :class:`~concurrent.futures.Executor` is passed instead, it will be used as-is.
+
+
+    The behavior for different combinations of `executor` and `ncpu` is:
+    +-----------------+--------+------------------------+
+    | `executor`      | `ncpu` | result                 |
+    +=================+========+========================+
+    | existing object | any    | uses existing executor |
+    +-----------------+--------+------------------------+
+    | None	      | 1      | uses single-processor  |
+    +-----------------+--------+------------------------+
+    | None	      | None   | creates default        |
+    |                 |        | executor with          |
+    |                 |        | ``cpu_count`` workers  |
+    +-----------------+--------+------------------------+
+    | None	      | >1     | creates default        |
+    |                 |        | executor with desired  |
+    |                 |        | number of workers      |
+    +-----------------+--------+------------------------+
+    """
+
+    if isinstance(executor, concurrent.futures.Executor):
```

Comment:
Python prefers "duck typing" - if it walks like a duck and quacks like a duck, it's a duck - to actual `isinstance`. Better to just ask if `exceutor is not None`.